### PR TITLE
Fix var/const/let variable use before declaration

### LINKED ISSUE #4101: `var` use before defined incorrect, `let`/`const` TDZ errors masked by tree shaking
- Rollup Version: rollup v2.50.1
- Operating System (or Browser): n/a
- Node Version (if applicable): v16.1.0

### Repro

```
$ cat misc.js
(function bug1() {
    var first = state();
    var on = true;
    var obj = { state: state };
    console.log(first, obj.state());
    function state() {
        return on ? "ON" : "OFF";
    }
})();
(function bug2() {
    var a = "PASS";
    for (var b = 2; --b >= 0; ) {
        (function() {
            var c = function() {
                return 1;
            }(c && (a = "FAIL"));
        })();
    }
    console.log(a);
})();
```

### Expected Behavior

```
$ cat misc.js | node
OFF ON
PASS
```

### Actual Behavior

```
$ cat misc.js | rollup --silent | node
ON ON
FAIL
```
```
$ cat misc.js | rollup --silent
(function bug1() {
    var first = state();
    var obj = { state: state };
    console.log(first, obj.state());
    function state() {
        return "ON" ;
    }
})();
(function bug2() {
    var a = "PASS";
    for (var b = 2; --b >= 0; ) {
        (function() {
            (function() {
                return 1;
            })((a = "FAIL"));
        })();
    }
    console.log(a);
})();
```

### PR DESCRIPTION
**Pull request by kzc**
* Also retains TDZ violations present in input
* Enabled by default when treeshake is active
* Low overhead - uses existing treeshaking simulated execution to mark declarations as reached

This PR contains:
- [x] bugfix
- [ ] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?
- [x] yes (*bugfixes and features will not be merged without tests*)
- [ ] no

Breaking Changes?
- [ ] yes (*breaking changes will not be merged unless absolutely necessary*)
- [x] no

List any relevant issue numbers:

resolves #4101

### Description

Retains code that reads const/let/var variables before their declarations to preserve the original behavior of the rollup input. This PR handles all known issues in #4101.

### CODE REVIEW

**Comment by kzc** [comment]
Not much can be done about this...

* codecov/project — 98.23% (-0.05%)

* Tests / Node 12 (Windows):

npm WARN tarball tarball data for typescript@4.3.2 (sha512-zZ4hShnmnoVnAHpVHWpTcxdv7dWP60S2FsydQLV8V5PbS3FifjWFFRiHSWpDJahly88PRyV5teTSLoq4eG7mKw==) seems to be corrupted.

**Comment by kzc** [comment]
I couldn't examine the Codecov report in any event:
```
Unauthorized

Github API rate limit error: Forbidden Check on Codecov's status or see our docs for common support.

Error 403
```

**Comment by kzc** [comment]
@lukastaegert This PR successfully runs the tests from #4149 without `treeshake.correctVarBeforeDeclaration` being set.

`codecov/project — 98.23% (-0.05%)` cannot be resolved without adding unrelated coverage tests outside the scope of this PR.

File: src/ast/nodes/Identifier.ts [hunk 1]
```diff
@@ -15,13 +15,29 @@ import LocalVariable from '../variables/LocalVariable';
...
+
+const variableKinds = {
+	__proto__: null,
+	const: true,
+	let: true,
+	var: true
+};
+
 export default class Identifier extends NodeBase implements PatternNode {
+	TDZ: boolean | undefined = undefined;
```
Comment (lukastaegert) [inline src/ast/nodes/Identifier.ts:40]: following our naming conventions, this should not start with a capital letter, so probably rather `tdz`, or maybe better `isTDZAccess` to clarify it is a boolean.

**Review by lukastaegert — CHANGES_REQUESTED** [review]
Great work, this is really looking very promising. I added some notes as I think especially the pattern handling could be improved.
Also there is one thing bothering me, but maybe this is not actually an issue:
Technically, there can be many tree-shaking passes. The reason is that e.g. when a declaration is first encountered, it is not included as it is unused at that point. Then later we use the declaration, which will include the corresponding variable and trigger another pass, at which point the declaration is then included.
During the second pass, however, the `initReached` values will be wrong as they retain the values at the end of the first pass.
The reason this is likely not an issue is that the TDZ values are cached anyway the first time an identifier is encountered. But I am not sure I am overlooking something, will need to think a little more about this. But if it should turn out to be an issue anyway, it would be solvable e.g. by putting a `Set` on the `HasEffectsContext` and `InclusionContext` to track which variables are initialized and use a fresh `Set` for each pass. But as I said, maybe we really do not have an issue here.

File: src/ast/nodes/Identifier.ts [hunk 2]
```diff
@@ -15,13 +15,29 @@ import LocalVariable from '../variables/LocalVariable';
...
+	FunctionExpression: true,
+	ObjectExpression: true
+};
+
+const variableKinds = {
+	__proto__: null,
+	const: true,
+	let: true,
+	var: true
+};
```
[the remark this answers is not in the document]
Comment (kzc) [reply src/ast/nodes/Identifier.ts:37]: Good point. I missed that.

File: src/ast/nodes/ArrayPattern.ts [hunk 3]
```diff
@@ -29,6 +29,9 @@ export default class ArrayPattern extends NodeBase implements PatternNode {
 				variables.push(...element.declare(kind, UNKNOWN_EXPRESSION));
 			}
 		}
+		variables.forEach(v => {
+			v.initReached = true;
+		});
```
[the remark this answers is not in the document]
Comment (kzc) [reply src/ast/nodes/ArrayPattern.ts:34]: Destructuring could definitely use more work.

File: src/ast/nodes/Identifier.ts [hunk 4]
```diff
@@ -188,6 +206,60 @@ export default class Identifier extends NodeBase implements PatternNode {
...
+			(init = (this.variable as any).init) &&
+			init !== UNDEFINED_EXPRESSION &&
+			(init_parent = (init as any).parent) &&
+			init_parent.type == 'VariableDeclarator' &&
+			init_parent.id.variable === this.variable &&
+			!(init_parent.init.type in tdzInitTypesToIgnore) &&
+			// code position comparisons must be in the same context
+			this.context === init_parent.id.context &&
+			this.start >= init.start &&
+			this.start < init.end
```
[the remark this answers is not in the document]
Comment (kzc) [reply src/ast/nodes/Identifier.ts:253]: That's an interesting case that I hadn't considered.

File: src/ast/nodes/Identifier.ts [hunk 5]
```diff
@@ -115,6 +130,9 @@ export default class Identifier extends NodeBase implements PatternNode {
 
 	hasEffects(): boolean {
 		if (!this.deoptimized) this.applyDeoptimizations();
+		if (this.isPossibleTDZ()) {
+			return true;
```
[the remark this answers is not in the document]
Comment (kzc) [reply src/ast/nodes/Identifier.ts:134]: I have to disagree with you on this point. This PR demonstrates that TDZ and var-use-before-decl are the same problem - they just result in different sorts of errors. 

I believe that case is fixable with some work. It is notable that it can be resolved with another pass:
```
$ echo 'var u = x; var x = 1;' | rollup --silent
x; var x = 1;
$ echo 'var u = x; var x = 1;' | rollup --silent | rollup --silent

$ 
```
In an earlier unrelated issue I examined how rollup might accomplish this without a reparse but got hung up on unexpected mutable behavior as result of repeated calls involving HasEffectsContext.

**Comment by kzc** [comment]
@lukastaegert Please feel free to modify this PR to your liking. Unfortunately I don't have the time to work on it further. I believe you have a good understanding of how it works and the remaining unresolved issues that are carried over from the previous rollup implementation. I would ask that whatever form this PR takes, that its functionality is the default treeshaking behavior and not be put behind an optional flag.

File: src/ast/nodes/Identifier.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (lukastaegert) [reply src/ast/nodes/Identifier.ts:134]: Ah, I see. First I though that the deoptimization via `getLiteralValueAtPath` is enough, but indeed it extends to all other functions as well. Basically what should happen is that before declaration, any identifier should behave like "UNKNOWN_EXPRESSION". Maybe the simple fix would be to replace `this.variable` with some `this.getValue` that returns either `this.variable` or `UNKNOWN_EXPRESSION`, or maybe entirely replace `this.variable` with `UNKNOWN_EXPRESSION` for TDZ values. Will need to think about this.
In any case, this would again make the second pass unnecessary as `UNKNOWN_EXPRESSION` does not have a side effect when accessed.
Also I think the access side effect for TDZ variables should be moved from `hasEffects` to `hasEffectsWhenAccessedAtPath`.

**Comment by kzc** [comment]
I think it's worth noting why the regression in `test/form/samples/simplify-return-expression/main.js` happened in its original form with this PR. If that source module were to be loaded by a JS engine it would not perform any work other than setting some top level variables (test, foo, bar, getBuild) to their respective arrow functions, and then setting BUILD to a constant literal - no functions are called. But because rollup `include()`d the exported function `test` in order of appearance in the source file, this PR believes that accessing BUILD before its declaration is a TDZ violation, which pessimizes and retains that variable. The reason that BUILD does not result in a TDZ in practise in a JS engine is due to the fact that it is a deferred operation, since the exported function in question, `test`, is not actually run. The distinction is subtle, but it makes a difference. You will notice that the IIFE version of the same code that calls `test()` at the end of IIFE does in fact drop BUILD. Also, had that original module been imported and `test` called then BUILD would also be dropped. However, it may be possible to tweak the `include` algorithm to defer the inclusion of the exported functions to the end of the module, thus not triggering the TDZ, and achieve the tree shaking of BUILD.

File: src/ast/nodes/ArrayPattern.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (lukastaegert) [reply src/ast/nodes/ArrayPattern.ts]: I pushed a fix for the destructuring as outlined

File: src/ast/nodes/Identifier.ts [hunk 6]
```diff
@@ -15,18 +15,34 @@ import LocalVariable from '../variables/LocalVariable';
...
+	var: true
+};
+
 export default class Identifier extends NodeBase implements PatternNode {
 	name!: string;
 	type!: NodeType.tIdentifier;
-
 	variable: Variable | null = null;
 	protected deoptimized = false;
+	private isTDZAccess = false;
```
Comment (kzc) [inline src/ast/nodes/Identifier.ts:45]: This is a tri-state property. For the caching to work in:

https://github.com/rollup/rollup/blob/41d5fde334c814d10d913fab1f9ed8ed701dc191/src/ast/nodes/Identifier.ts#L213-L215

it needs to be:
```
private isTDZAccess: boolean | undefined = undefined;
```
which resolves the 11 npm test failures in 41d5fde334c814d10d913fab1f9ed8ed701dc191.

Comment (lukastaegert) [reply src/ast/nodes/Identifier.ts:45]: Ah, that makes a lot of sense

**Comment by lukastaegert** [comment]
> The reason that BUILD does not result in a TDZ in practise in a JS engine is due to the fact that it is a deferred operation, since the exported function in question, test, is not actually run. The distinction is subtle, but it makes a difference

This analysis is spot on. I added a fix that defers including entry exports after the first tree-shaking pass that indeed fixes this.

**Comment by kzc** [comment]
Good stuff - and such a simple fix.

What's left to do? This PR already fixes many longstanding tree shaking and ES correctness issues.

File: src/ast/nodes/Identifier.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (lukastaegert) [reply src/ast/nodes/Identifier.ts:131]: This is fixed now

**Comment by lukastaegert** [comment]
I want to have a look at this one https://github.com/rollup/rollup/pull/4148#discussion_r659474497
and also deprecate the correctVarValueBeforeDeclaration option properly.

File: src/ast/nodes/Identifier.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (lukastaegert) [reply src/ast/nodes/Identifier.ts]: Fixed.

**Comment by lukastaegert** [comment]
This is mostly done, I am currently looking into some edge cases. So looking at the coverage misses, there is indeed an interesting deoptimization behind this: https://rollupjs.org/repl/?pr=4148&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmZ1bmN0aW9uJTIwZm9vKCklMjAlN0IlNUNuJTVDdGNvbnN0JTIwcmVzdWx0JTIwJTNEJTIwZmFsc2UlM0IlNUNuJTVDdHJldHVybiUyMHJlc3VsdCUzQiU1Q24lN0QlM0IlNUNuJTVDbiUyRiUyRiUyMFRoZSUyMGxpbmUlMjBiZWxvdyUyMGlzJTIwZGVvcHRpbWl6aW5nJTIwZXZlcnl0aGluZyU1Q25jb25zb2xlLmxvZyhmb28oKSUyMCU3QyU3QyUyMHRydWUpJTNCJTVDbiU1Q25pZiUyMChmb28oKSUyMCU3QyU3QyUyMHRydWUpJTIwY29uc29sZS5sb2coJ3JldGFpbmVkJyklM0IlNUNuZWxzZSUyMGNvbnNvbGUubG9nKCdzaG91bGQlMjBiZSUyMHJlbW92ZWQlMjBidXQlMjBpcyUyMG5vdCcpJTNCJTVDbiU1Q25mdW5jdGlvbiUyMGJhcigpJTIwJTdCJTVDbiU1Q3Rjb25zdCUyMHJlc3VsdCUyMCUzRCUyMGZhbHNlJTNCJTVDbiU1Q3RyZXR1cm4lMjByZXN1bHQlM0IlNUNuJTdEJTNCJTVDbiU1Q25pZiUyMChiYXIoKSUyMCU3QyU3QyUyMHRydWUpJTIwY29uc29sZS5sb2coJ3JldGFpbmVkJyklM0IlNUNuZWxzZSUyMGNvbnNvbGUubG9nKCdyZW1vdmVkJyklM0IlMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiUyQyUyMm5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCU3RCUyQyUyMmV4YW1wbGUlMjIlM0FudWxsJTdE

Without the patch, the logical expression would have been simplified for both foo and bar. But with the patch, it first tries to deoptimize the arguments of the console.log call, and during this deoptimization it thinks that the variable `result` in the function call is in a TDZ position (because the function was not checked at this point; to be precise, this happens when the logical expression checks for the literal value of its left side). This value is then stored with the identifier.

**Comment by lukastaegert** [comment]
And here is an actual bug which stems from the fact that during the first traversal of foo, value was not accessed in TDZ position. Then the variable is marked as "declared" and only during the second pass, the TDZ access is reached, at which point it is no longer recognized as such: https://rollupjs.org/repl/?pr=4148&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmxldCUyMHRkekFjY2VzcyUyMCUzRCUyMGZhbHNlJTNCJTVDbiU1Q25mdW5jdGlvbiUyMGZvbygpJTIwJTdCJTVDbiU1Q3RpZiUyMCh0ZHpBY2Nlc3MpJTIwJTdCJTVDbiU1Q3QlMjAlMjB2YWx1ZSUzQiU1Q24lNUN0JTdEJTVDbiU1Q3RsZXQlMjB2YWx1ZSU1Q24lN0QlM0IlNUNuJTVDbiUyRiUyRiUyMHJlbW92ZSUyMHRoZSUyMGZpcnN0JTIwY2FsbCUyMHRvJTIwZml4JTIwdGhlJTIwaXNzdWUlNUNuZm9vKCklM0IlNUNuJTVDbnRkekFjY2VzcyUyMCUzRCUyMHRydWUlM0IlNUNuZm9vKCklM0IlMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiUyQyUyMm5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCU3RCUyQyUyMmV4YW1wbGUlMjIlM0FudWxsJTdE

I am currently wondering if we do not want to refine the approach, maybe even going as far as adding the TDZ flags in a separate step.

**Comment by kzc** [comment]
At least the bug example above is not a regression. The same erroneous empty output is produced with rollup v2.52.7 as well. Something else appears to be amiss. 

This PR may not handle every TDZ scenario correctly, but it's a big improvement in any case. Detecting all TDZ issues is not possible without actually executing the code.

I'll take a look at these examples tomorrow if you haven't already figured out a solution by that time.

**Comment by kzc** [comment]
The following patch applied to the latest PR commit will fix the bug in https://github.com/rollup/rollup/pull/4148#issuecomment-873345611 and pass `npm test`:

```patch
--- a/src/ast/nodes/Identifier.ts
+++ b/src/ast/nodes/Identifier.ts
@@ -239,6 +239,20 @@ export default class Identifier extends NodeBase implements PatternNode {
                        return (this.isTDZAccess = false);
                }
 
+               var decl_id;
+               if (
+                       this.variable.kind === 'let' &&
+                       this.variable.declarations &&
+                       this.variable.declarations.length === 1 &&
+                       (decl_id = this.variable.declarations[0] as any) &&
+                       this.start < decl_id.start &&
+                       closestParentFunctionOrProgram(this) === closestParentFunctionOrProgram(decl_id)
+               ) {
+                       // a `let` variable accessed before its declaration
+                       // in the same function or at top level of module
+                       return (this.isTDZAccess = true);
+               }
+
                if (!this.variable.initReached) {
                        // Either a const/let TDZ violation or
                        // var use before declaration was encountered.
@@ -248,3 +262,11 @@ export default class Identifier extends NodeBase implements PatternNode {
                return (this.isTDZAccess = false);
        }
 }
+
+function closestParentFunctionOrProgram(node: any): any {
+       while (node && !/^Program|Function/.test(node.type)) {
+               node = node.parent;
+       }
+       // One of: ArrowFunctionExpression, FunctionDeclaration, FunctionExpression or Program.
+       return node;
+}
```

I have not investigated the working but suboptimal test case in https://github.com/rollup/rollup/pull/4148#issuecomment-873345209. There are bound to be edge cases, but as long as they are infrequent and produce the correct result they can be tolerated until a future PR fix can be devised.

File: docs/999-big-list-of-options.md [hunk 7]
```diff
@@ -1760,6 +1739,13 @@ Default: `import`
 
 This will rename the dynamic import function to the chosen name when outputting ES bundles. This is useful for generating code that uses a dynamic import polyfill such as [this one](https://github.com/uupaa/dynamic-import-polyfill).
 
+**treeshake.correctVarValueBeforeDeclaration**<br>
+Type: `boolean`<br>
+CLI: `--treeshake.correctVarValueBeforeDeclaration`/`--no-treeshake.correctVarValueBeforeDeclaration`<br>
+Default: `false`
+
+This was temporarily an option to deoptimize the values of all `var` declarations to handle accessing a variable before it is declared. This option no longer has an effect as there is now an improved logic in place to automatically detect these situations. The option will be removed entirely in future Rollup versions.
```
Comment (kzc) [inline docs/999-big-list-of-options.md:1747]: s/an improved logic/improved logic/

**Comment by kzc** [comment]
The following patch fixes the depot in https://github.com/rollup/rollup/pull/4148#issuecomment-873345209 without any `npm test` failures:

```patch
--- a/src/ast/nodes/CallExpression.ts
+++ b/src/ast/nodes/CallExpression.ts
@@ -178,15 +178,18 @@ export default class CallExpression extends NodeBase implements DeoptimizableEnt
        hasEffects(context: HasEffectsContext): boolean {
-               if (!this.deoptimized) this.applyDeoptimizations();
-               for (const argument of this.arguments) {
-                       if (argument.hasEffects(context)) return true;
+               try {
+                       for (const argument of this.arguments) {
+                               if (argument.hasEffects(context)) return true;
+                       }
+                       if (
+                               (this.context.options.treeshake as NormalizedTreeshakingOptions).annotations &&
+                               this.annotations
+                       )
+                               return false;
+                       return (
+                               this.callee.hasEffects(context) ||
+                               this.callee.hasEffectsWhenCalledAtPath(EMPTY_PATH, this.callOptions, context)
+                       );
+               } finally {
+                       if (!this.deoptimized) this.applyDeoptimizations();
                }
-               if (
-                       (this.context.options.treeshake as NormalizedTreeshakingOptions).annotations &&
-                       this.annotations
-               )
-                       return false;
-               return (
-                       this.callee.hasEffects(context) ||
-                       this.callee.hasEffectsWhenCalledAtPath(EMPTY_PATH, this.callOptions, context)
-               );
        }
```

I used try/finally because it was the easiest way to defer `this.applyDeoptimizations()` without thinking of the `return` statement code flow.  It can be manually expanded of course with the risk of introducing mechanical errors - if you think it's more efficiently handled by the V8 engine.

Other AST classes should be examined for similar suboptimal code due to child node `hasEffects*` and `applyDeoptimizations` ordering. NewExpression would probably have the same issue.

**Comment by kzc** [comment]
This is the same patch as above but it's easier to read using `git diff -w`:

```patch
--- a/src/ast/nodes/CallExpression.ts
+++ b/src/ast/nodes/CallExpression.ts
@@ -176,7 +176,7 @@ export default class CallExpression extends NodeBase implements DeoptimizableEnt
        }
 
        hasEffects(context: HasEffectsContext): boolean {
-               if (!this.deoptimized) this.applyDeoptimizations();
+               try {
                        for (const argument of this.arguments) {
                                if (argument.hasEffects(context)) return true;
                        }
@@ -189,6 +189,9 @@ export default class CallExpression extends NodeBase implements DeoptimizableEnt
                                this.callee.hasEffects(context) ||
                                this.callee.hasEffectsWhenCalledAtPath(EMPTY_PATH, this.callOptions, context)
                        );
+               } finally {
+                       if (!this.deoptimized) this.applyDeoptimizations();
+               }
        }
 
        hasEffectsWhenAccessedAtPath(path: ObjectPath, context: HasEffectsContext): boolean {
```

File: src/ast/nodes/Identifier.ts [hunk 8]
```diff
@@ -46,14 +62,9 @@ export default class Identifier extends NodeBase implements PatternNode {
...
 	declare(kind: string, init: ExpressionEntity): LocalVariable[] {
 		let variable: LocalVariable;
-		const { treeshake } = this.context.options;
 		switch (kind) {
 			case 'var':
 				variable = this.scope.addDeclaration(this, this.context, init, true);
-				if (treeshake && treeshake.correctVarValueBeforeDeclaration) {
-					// Necessary to make sure the init is deoptimized. We cannot call deoptimizePath here.
-					this.scope.addDeclaration(this, this.context, UNDEFINED_EXPRESSION, true);
-				}
```
[the remark this answers is not in the document]
Comment (lukastaegert) [reply src/ast/nodes/Identifier.ts]: Keeping this for now as there are still some edge cases to be resolved

**Review by lukastaegert — APPROVED** [review]
I added your fix for the CallExpression regression but did not add the other fix as I find it does not really fix the root cause but just my constructed example, and there could be quite a few variations that are not yet fixed. I would rather go for a separate approach to these ones. Which is why I undeprecated the correctVarValueBeforeDeclaration for now.
Still, I think this could be released for now as a definitive improvement.

**Comment by kzc** [comment]
> but did not add the other fix as I find it does not really fix the root cause but just my constructed example, and there could be quite a few variations that are not yet fixed

That's unfortunate.  It's a very simple and reliable approach to fix such bugs - any use of a `let` variable before its declaration within the exact same function (or both in top level) and not crossing function boundaries ought to always be a TDZ.

Can you post some examples of variations that are not fixed by https://github.com/rollup/rollup/pull/4148#issuecomment-873450839 ?

**Comment by kzc** [comment]
> Which is why I undeprecated the correctVarValueBeforeDeclaration for now.

I suppose there's no harm in keeping this around as a non-default treeshake option, but as of this latest PR, I'm not aware of any rollup input producing a wrong result for var-use-before-declaration. The [patch not applied](https://github.com/rollup/rollup/pull/4148#issuecomment-873450839) is a [`let` TDZ fix](https://github.com/rollup/rollup/pull/4148#issuecomment-873345611) - not related to `var`.

**Comment by lukastaegert** [comment]
Can you post some examples of variations that are not fixed?
* Doing it in a function call with deferred declaration: https://rollupjs.org/repl/?pr=4148&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmxldCUyMHRkekFjY2VzcyUyMCUzRCUyMGZhbHNlJTNCJTVDbiU1Q25mdW5jdGlvbiUyMGZvbygpJTIwJTdCJTVDbiU1Q3RpZiUyMCh0ZHpBY2Nlc3MpJTIwJTdCJTVDbiU1Q3QlMjAlMjBhY2Nlc3MoKSUzQiU1Q24lNUN0JTdEJTVDbiU1Q3RsZXQlMjB2YWx1ZSU1Q24lNUN0JTVDbiU1Q3RmdW5jdGlvbiUyMGFjY2VzcygpJTIwJTdCJTVDbiU1Q3QlMjAlMjB2YWx1ZSUzQiU1Q24lNUN0JTdEJTVDbiU3RCUzQiU1Q24lNUNuJTJGJTJGJTIwcmVtb3ZlJTIwdGhlJTIwZmlyc3QlMjBjYWxsJTIwdG8lMjBmaXglMjB0aGUlMjBpc3N1ZSU1Q25mb28oKSUzQiU1Q24lNUNudGR6QWNjZXNzJTIwJTNEJTIwdHJ1ZSUzQiU1Q25mb28oKSUzQiUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzJTIyJTJDJTIybmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=
* You have the same issue with `var`: https://rollupjs.org/repl/?pr=4148&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmxldCUyMHRkekFjY2VzcyUyMCUzRCUyMGZhbHNlJTNCJTVDbiU1Q25mdW5jdGlvbiUyMGZvbygpJTIwJTdCJTVDbiU1Q3RpZiUyMCh0ZHpBY2Nlc3MpJTIwJTdCJTVDbiU1Q3QlNUN0aWYlMjAoIXZhbHVlKSUyMGNvbnNvbGUubG9nKCdzaG91bGQlMjBiZSUyMHJldGFpbmVkJyklM0IlNUNuJTVDdCU3RCU1Q24lNUN0dmFyJTIwdmFsdWUlMjAlM0QlMjB0cnVlJTNCJTVDbiU3RCUzQiU1Q24lNUNuJTJGJTJGJTIwcmVtb3ZlJTIwdGhlJTIwZmlyc3QlMjBjYWxsJTIwdG8lMjBmaXglMjB0aGUlMjBpc3N1ZSU1Q25mb28oKSUzQiU1Q24lNUNudGR6QWNjZXNzJTIwJTNEJTIwdHJ1ZSUzQiU1Q25mb28oKSUzQiUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzJTIyJTJDJTIybmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=
* On the other hand, it would needlessly deoptimize this access via arrow function https://rollupjs.org/repl/?pr=4148&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmZ1bmN0aW9uJTIwZm9vKCklMjAlN0IlNUNuJTVDdGNvbnN0JTIwYWNjZXNzJTIwJTNEJTIwKCklMjAlM0QlM0UlMjB2YWx1ZSUzQiU1Q24lNUN0bGV0JTIwdmFsdWUlM0IlNUNuJTVDdGFjY2VzcygpJTNCJTVDbiU3RCUzQiU1Q24lNUNuZm9vKCklM0IlNUNuJTIyJTJDJTIyaXNFbnRyeSUyMiUzQXRydWUlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==

I think we should fix these issues, but I also think we should reflect a little deeper on the how. As we already fixed a lot of issues, I will release this now.
My feeling is that a "proper" solution would not assign fixed flags to variables, but rather tracks on the context what variables are initialized and also "forgets" this at the end of a function call. But of course, such a solution would be much more involved. On the other hand, this might also allow us to remove the blanket deoptimization we currently have for `var` declarations in nested block scopes:
https://rollupjs.org/repl/?pr=4148&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMiU3QiU1Q24lMjAlMjB2YXIlMjBuZXN0ZWQlMjAlM0QlMjBmYWxzZSUzQiU1Q24lNUN0aWYlMjAobmVzdGVkKSUyMGNvbnNvbGUubG9nKCd0aGlzJTIwY291bGQlMjBiZSUyMHJlbW92ZWQnKSUzQiU1Q24lN0QlMjIlMkMlMjJpc0VudHJ5JTIyJTNBdHJ1ZSU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiUyQyUyMm5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCU3RCUyQyUyMmV4YW1wbGUlMjIlM0FudWxsJTdE
This exists because otherwise, we might fail situations like this:
https://rollupjs.org/repl/?pr=4148&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMnN3aXRjaCUyMCh1bmtub3duKSUyMCU3QiU1Q24lNUN0Y2FzZSUyMGZpcnN0JTNBJTVDbiU1Q3QlNUN0dmFyJTIweCUyMCUzRCUyMHRydWUlM0IlNUNuJTVDdGNhc2UlMjBzZWNvbmQlM0ElNUNuJTVDdCU1Q3RpZiUyMCh4KSUyMGNvbnNvbGUubG9nKCdvbmx5JTIwaWYlMjAlNUMlMjJmaXJzdCU1QyUyMiUyMHdhcyUyMHVzZWQlMkMlMjBibGFua2V0JTIwZGVvcHRpbWl6YXRpb24lMjBzYXZlZCUyMHVzJTIwaGVyZScpJTNCJTVDbiU3RCUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzJTIyJTJDJTIybmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=
And you can also constructor similar situations with if statements.

**Comment by kzc** [comment]
This PR is a net improvement over the previous release, which is good. But the thing is that the PR only exists because some of rollup's optimization assumptions regarding `init` expressions were incorrect. So future fixes would involve work from both sides of the issue - questioning whether some optimizations are tracking variable state changes correctly in the first place, and adding variable deoptimizations when it cannot be proven that an access can be safely elided.
